### PR TITLE
Update readme

### PR DESCRIPTION
**Pull request by alexdavid**
@charlierudolph @kevgo 

Update the readme to document all `react-e` features and compare them with `react.createElement` and JSX

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,47 +1,161 @@
...
+&lt;Button
+  className={styles.foo + ' ' + styles[someVar]}
+&gt;&lt;/Button&gt;
+    </pre></td>
+  </tr>
+</table>
+
+
+## Installing
+`npm install --save react-e`
```
Comment (charlierudolph) [inline README.md]: In other modules, I feel like I usually see install at the top above usage. Why did you move it to the bottom.

File: README.md [hunk 2]
```diff
@@ -1,47 +1,161 @@
...
-</div>
-```
+e('.foo', {className: someVar})
+    </pre></td>
+    <td><pre>
+styles = require('./stylesheet.css')
+
+createElement('div', {
+  className: styles.foo + ' ' + styles[someVar]
+})
```
Comment (charlierudolph) [inline README.md]: I think this and the jsx example needs to use `classnames.bind`. Thats an important part as otherwise its not the same.

File: README.md (hunk 1 above)
Comment (alexdavid) [reply README.md]: I felt like it's less important, most people will see `require 'react-e'` at the top and assume it's `npm install react-e`. But I don't feel too strongly about it.

File: README.md (hunk 2 above)
Comment (alexdavid) [reply README.md]: I wanted to document the actual behavior in case users weren't familiar with `classnames.bind`. It ends up being equivalent, it may not be a fair comparison here though

Comment (alexdavid) [reply README.md]: After thinking about this more, I agree

Comment (charlierudolph) [reply README.md]: `styles[someVar]` isn't equivalent though. Assuming `someVar` is a string, I believe its actually `styles[someVar] or "#{someVar}"` in the case styles doesn't define a mapping for that string. classnames also deals with objects, arrays, and if `someVar` is undefined.

Comment (alexdavid) [reply README.md]: Ah yeah, good point

**Comment by charlierudolph** [comment]
All the code blocks in the features section need syntax highlighting. Looks great Alex! This is awesome.

File: README.md [hunk 3]
```diff
@@ -1,47 +1,176 @@
...
+## Installing
+`npm install --save react-e`
+
+
+## Features
+
+React-e is a drop in replacement for `react.createElement`, but it provides syntactic advantages:
+
+#### `props` can be omitted even if you have children:
+<table>
```
Comment (kevgo) [inline README.md:35]: This table is too wide to display. Maybe organize it vertically?
- 2 columns, 3 rows
- left column contains the headers
- right column contains the code examples

File: README.md [hunk 4]
```diff
@@ -1,47 +1,176 @@
...
+    </pre></td>
+    <td><pre lang="jsx">
+&lt;div&gt;
+  &lt;span&gt;Hello world&lt;/span&gt;
+&lt;/div&gt;
+    </pre></td>
+  </tr>
+</table>
+
+#### static classnames can be provided inline using jade syntax:
```
Comment (kevgo) [inline README.md]: Jade has been renamed to pug

File: README.md [hunk 5]
```diff
@@ -1,47 +1,176 @@
...
+<table>
+  <tr>
+    <th>react-e example</th>
```
Comment (kevgo) [inline README.md]: Thoughts about naming this "react-e version"? I think you are providing 3 different versions of the same example here.

File: README.md [hunk 6]
```diff
@@ -1,47 +1,176 @@
...
-  e 'span.text', className: active: true, 'Some text'
-```
-Is equivalent to:
-```jsx
-<div className="local1">
-  <button className="local2">Click me</button>
-  <span className="local3 active">Some text</span>
-</div>
-```
+e('.foo', {className: someVar})
```
Comment (kevgo) [inline README.md]: Where does `someVar` come from here? How is it related to the CSS required earlier?

**Comment by alexdavid** [comment]
👍 Didn't know how to add syntax highlighting to `<pre>` tags in markdown but figured it out!

File: README.md [hunk 7]
```diff
@@ -1,47 +1,180 @@
...
-```
-Is equivalent to:
-```jsx
-<div className="local1">
-  <button className="local2">Click me</button>
-  <span className="local3 active">Some text</span>
-</div>
-```
+e('.foo', {
+  className: someVar
```
Comment (kevgo) [inline README.md]: Where does `someVar` come from? How is it related to the CSS required earlier?

File: README.md (hunk 3 above)
Comment (alexdavid) [reply README.md:35]: 👍 I like this idea

File: README.md (hunk 5 above)
Comment (alexdavid) [reply README.md]: What do you think about just calling it "react-e"?

File: README.md (hunk 6 above)
Comment (alexdavid) [reply README.md]: It would be a variable containing some css classname. Would a better variable name or a `someVar = "some-class"` at the top of the file make it more clear?

File: README.md (hunk 5 above)
Comment (kevgo) [reply README.md]: works for me. short and sweet

File: README.md (hunk 6 above)
Comment (kevgo) [reply README.md]: Maybe provide an inline version?

```
e('.foo', {className: 'my-css-class'})
```

I still don't see the connection to loading the CSS file earlier, though

File: README.md (hunk 3 above)
Comment (kevgo) [reply README.md:35]: The formatting is too blocky. Maybe try `<code>` instead of `<pre>`?

File: README.md (hunk 6 above)
Comment (alexdavid) [reply README.md]: This is a css modules thing. It's documented in the link I provide in the readme. Basically it converts your classes to globally-unique classnames and provides a hash that maps your human-readable classnames to the actual ones

```coffeescript
styles = {
  'my-css-class': 'f6aUfbnuw_bn'
}
```

File: README.md (hunk 3 above)
Comment (alexdavid) [reply README.md:35]: `<code>` doesn't seem to work with multi-line code blocks

File: README.md (hunk 6 above)
Comment (alexdavid) [reply README.md]: I think inlining the string there makes a worse example becuase otherwise you'd just use `e('.foo.my-css-class')`. I want to document the actual use-case which is when you have a variable with a class name in it

File: README.md [hunk 8]
```diff
@@ -1,47 +1,197 @@
...
+React-e is a drop in replacement for `react.createElement`, but it provides syntactic advantages:
+
+#### `props` can be omitted even if you have children:
+<table>
+  <tr>
+    <th>react-e</th>
+    <td><pre lang="javascript">
+e('div',
+  e('span', 'Hello world'))
+    </pre></td>
```
Comment (charlierudolph) [inline README.md]: If you move the `</pre></td>` to end of the previous line we remove the extra new line at the end of the code block

File: README.md (hunk 3 above)
Comment (kevgo) [reply README.md:35]: 😿

File: README.md (hunk 6 above)
Comment (kevgo) [reply README.md]: Maybe providing the CSS file (containing a style called `someVar`) above the table helps make it more understandable where `someVar` is coming from? If I understand this correctly...

Comment (alexdavid) [reply README.md]: it works like this, but that functionality isn't provided by this library and is documented in the link I provide so I'm not sure how much sense it makes to bloat up the readme with this full example:


```stylus
// button.styl

.button
  cursor: pointer

.primary
  background: blue
```

```coffeescript
# button.coffee
styles = require './button.styl'
e = require('react-e/bind')(styles)

Button = (props) ->
  e '.button', className: {primary: props.primary}, props.text
```

This compiles into something like:
```
.74g5f14nvb { cursor: pointer }
.b5aw5m9bs { background: blue }
```
and
```
styles = {
  button: '74g5f14nvb',
  primary: 'b5aw5m9bs'
}
```

Comment (charlierudolph) [reply README.md]: @alexdavid thoughts on just using the same classname from the previous example
```es6
className: ['foo', {active: this.state.active}]
```
instead of introducing a variable? The main part to me is just showing we the classnames/bind functionality built in and what gets passed in doesn't much matter

Comment (alexdavid) [reply README.md]: Yeah, I like `this.state.active`

**Review by charlierudolph — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -1,47 +1,185 @@
...
 
-```coffeescript
-styles = require 'index.css'
-###
-styles =
-  container: 'local1'
-  primary: 'local2'
-  text: 'local3'
-###
+## Installing
```
Comment (kevgo) [inline README.md]: Installation  ?

File: README.md [hunk 10]
```diff
@@ -1,47 +1,185 @@
 # react-e [![CircleCI](https://circleci.com/gh/Originate/react-e.svg?style=svg)](https://circleci.com/gh/Originate/react-e)
 
-React helper to generate elements
+JSX replacement that doesn't need to be compiled.
 
-## Installing
-`npm install --save react-e`
+## Basic Usage
```
Comment (kevgo) [inline README.md]: I would remove this caption. Without it, the introductory paragraph contains a one-sentence summary and a basic usage example. Makes the text flow a bit better imo.

File: README.md [hunk 11]
```diff
@@ -1,47 +1,185 @@
...
+e = require('react-e')
 
-```coffeescript
-e = require 'react-e'
-
-e '.container',
-  e 'button.primary', 'Click me'
-  e 'span.text', className: active: true, 'Some text'
+e(Modal,
+  e('.container',
```
Comment (kevgo) [inline README.md]: I still vote for using `$` instead of `e` here. It is a lot more intuitive than `e`. $ is in no way owned by jQuery, it's just a placeholder for "the DOM library you use".

File: README.md [hunk 12]
```diff
@@ -1,47 +1,185 @@
...
+```javascript
+e = require('react-e')
 
-```coffeescript
-e = require 'react-e'
-
-e '.container',
-  e 'button.primary', 'Click me'
-  e 'span.text', className: active: true, 'Some text'
+e(Modal,
```
Comment (kevgo) [inline README.md]: Where does `Model` come from here?

File: README.md (hunk 11 above)
Comment (alexdavid) [reply README.md]: I don't feel strongly either way because it's really up to the end user to decide what they want to use. `$` is fine with me. Do you have any objections @charlierudolph?

File: README.md (hunk 12 above)
Comment (alexdavid) [reply README.md]: I think it's obvious that it's another react component. I think adding extra mock requires to the top here just clutters up the example and distracts the user from what this library actually does

File: README.md (hunk 8 above)
Comment (alexdavid) [reply README.md]: 🏁

File: README.md (hunk 11 above)
Comment (kevgo) [reply README.md]: I'm still not a fan of the redundant markup. For example, to create a button, I need `e` and `Button`. Why can't `Button` be enough here? Even JSX does that with one piece of boilerplate. 

Here is how an optimized syntax would look like in Kevin's dream world inside his head:

```js
{$, Button, Modal, span} = require('react-e').domElements
// or $ = require('react-e') for the old way

Modal(
  $('.container',
    span('.text', {className: {active: this.state.active}}, 'Some text'),
    span({className: ['text', {active: this.state.active}]}, 'Some text'),  // alternative syntax
    Button({className: 'primary'}, 'Close')
  )
)
```

Just an idea  :)

File: README.md [hunk 13]
```diff
@@ -1,47 +1,185 @@
 # react-e [![CircleCI](https://circleci.com/gh/Originate/react-e.svg?style=svg)](https://circleci.com/gh/Originate/react-e)
 
-React helper to generate elements
+JSX replacement that doesn't need to be compiled.
```
Comment (kevgo) [inline README.md]: Let's brainstorm the tagline a bit more, since it is the thing that will sell this the best. 

What is the primary reason why you built react-e? I doubt that it was because you didn't want to compile stuff. We spoke about how the ideal solution is a compiled DSL that is JSX but better. I also doubt that you did it just to replace JSX, just because. 

What about `JSX alternative with better readability and productivity`?

File: README.md (hunk 11 above)
Comment (kevgo) [reply README.md]: Woops the last comment ended up in the same thread, sorry about that.

In reply to @alexdavid, you have a responsibility to define the best use case in the readme, since most people will follow your example. I also think `$` is more intuitive. jQuery had a very good reputation, most people would love to use something like it again.

Comment (alexdavid) [reply README.md]: The problem is Modal and Button come from the user and won't be exported from the library. Also every time I've destrucutred {div, ...} from react.DOM in the past that line has gotten horribly out of date and inevitably has a ton of unused tags destructured from it.

File: README.md (hunk 12 above)
Comment (kevgo) [reply README.md]: It wasn't obvious to me. `Model` sounds pretty generic. Maybe at least add a comment on top saying that you should load `Model` here? Or use `MyReactComponent` or something more self-describing?

File: README.md (hunk 13 above)
Comment (alexdavid) [reply README.md]: Primary reason is because JSX doesn't work well in coffeescript/livescript because of its need to be compiled. Projects like cjsx have failed because lack of tooling support. This library works well for non-babel react codebases.

File: README.md (hunk 12 above)
Comment (alexdavid) [reply README.md]: In react-land any PascalCase component is a user defined one

File: README.md (hunk 11 above)
Comment (kevgo) [reply README.md]: As long as there isn't a pedantic linter/type checker to complain about it, nobody has a problem. And if there is a pedantic linter to complain about it, it's easy to clean up.

Comment (alexdavid) [reply README.md]: I've just seen them turn into messes a lot. Especially keeping them in alphabetical order. After trying it on a few projects my recommendation is to not destructure DOM elements.

Comment (alexdavid) [reply README.md]: In your previous example it would have to be:
```
{$, span} = require('react-e').domElements
// or $ = require('react-e') for the old way

$(Modal,
  $('.container',
    span('.text', {className: {active: this.state.active}}, 'Some text'),
    span({className: ['text', {active: this.state.active}]}, 'Some text'),  // alternative syntax
    $(Button, {className: 'primary'}, 'Close')
  )
)
```
and since probably 90% of the elements used are going to be `div`s anyway, there's no real advantage to exposing the alternate syntax for the 10% of elements that aren't going to be divs

Comment (kevgo) [reply README.md]: Do you feel strongly enough about this to dictate that none of your users can ever use destructuring?

What about providing an easy way to use this library without destructuring (as you already do), and recommend that in the readme. And also allowing to use destructuring if people want to? I have found that a little bit of prudence goes a long way with such issues. I use destructuring on prelude-ls, often with 3-5 elements, and it's not a problem at all.

Comment (kevgo) [reply README.md]: Just an idea, we don't have to decide this here right now, since this PR is for the readme.

File: README.md (hunk 12 above)
Comment (kevgo) [reply README.md]: Still, I suggest using a name for the user-defined component that makes more clear that this is a user-defined component, and not some generic `Modal` thing that your library provides. Keep in mind that people who read this readme don't know what your library is, so they don't know that it doesn't provide a `Modal` thing.

File: README.md (hunk 11 above)
Comment (charlierudolph) [reply README.md]: I like `e` as it stands for element. In my project where we are using `e` we also have jquery in the project and use it in a couple places.

I also agree with @alexdavid in prefering this over destructing.

**Comment by alexdavid** [comment]
Shipping this for now since it's a huge improvement over the current readme. If you have other proposed changes we can build on it going forward
